Introduce system property extension

beatngu13 opened the pull request:
Introduces a Jupiter extension to clear (`@ClearSystemProperty`) or set (`@SetSystemProperty`) system properties for tests. Both annotations work on the test method and class level, are repeatable as well as combinable. Resolves #129.

```java
@ClearSystemProperty(key = "some property")
@Test
void clearSystemProperty() {
	assertThat(System.getProperty("some property")).isNull();
}

@SetSystemProperty(key = "some property", value = "new value")
@Test
void setSystemProperty() {
	assertThat(System.getProperty("some property")).isEqualTo("new value");
}
```

Resolves #129

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

smoyer64 commented on src/main/java/org/junitpioneer/jupiter/DefaultTimeZone.java:
```diff
@@ -22,7 +22,7 @@
  * returned by {@link java.util.TimeZone#getDefault()} for a test execution.
  *
  * <p>The {@link java.util.TimeZone} to set as the default {@code TimeZone} is
- * configured bei specifying the {@code TimeZone} ID as defined by
```
Is this change related to this PR? (don't get me wrong, I appreciate the effort to clean things up as their found).

smoyer64 replied:
If you wouldn't mind, and I promise I'll merge it immediately (are you signed up for HacktoberFest?).  As I noted above, I **do** appreciate people who help improve documentation!

beatngu13 replied:
There you go: https://github.com/junit-pioneer/junit-pioneer/pull/147 ✌️ Will shortly rebase and update this branch.

beatngu13 commented on src/main/java/org/junitpioneer/jupiter/ClearSystemProperty.java:
```diff
@@ -0,0 +1,42 @@
+	/**
+	 * The key of the system property to be cleared.
+	 */
+	String key();
```
I think I would go for the symmetric variant (i.e. with `key()`), but I have no strong opinion on that. If you prefer `value()`, I'm happy to change it.

nicolaiparlog replied:
I agree, symmetry over brevity.

beatngu13 commented:
Updated the docs, please let me know if there is anything else missing or you would like to change something.

I've also changed the copyright note, which was outdated. Do you know why Spotless complains about this change?

https://travis-ci.org/junit-pioneer/junit-pioneer/jobs/606463317#L291

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/ClearSystemProperties.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
```
Does it make sense to allow class-level annotations as well?

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/ClearSystemProperty.java:
```diff
@@ -0,0 +1,42 @@
+ * @since 0.4
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
```
(See above)

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/SystemPropertyExtension.java:
```diff
@@ -0,0 +1,86 @@
+
+	private void resetOriginalSystemProperties(final ExtensionContext context) {
+		final Properties backup = context.getStore(NAMESPACE).get(KEY, Properties.class);
+		System.setProperties(backup);
```
This restores all properties to the values they had before the test was run. If tests change system properties on purpose (sounds fishy, I know, but weirder things have happened), this extension overrides/negates that change. I think it would be better to just store/restore the properties used by the annotation.

To go further, maybe we should only restore system properties that have the value given to this extension? If a test sets any system property (even the one set for the test by this extension), we should not reset that. At least that's what my gut says atm. What do you think?

beatngu13 replied:
I now only store and restore those system properties that have been specified via the extension, not all per se. But I think it is fine to restore a system property in this situation:

>If a test sets any system property (even the one set for the test by this extension), we should not reset that.

I would argue that this is defined behavior as specified in the Javadoc. I also believe other extensions such as `DefaultLocaleExtension` and `DefaultTimeZoneExtension` work the same.

What do you say?

nicolaiparlog replied:
Makes sense, let's go with that. For now - maybe I'll come back to it in a separate issue.

nicolaiparlog commented on src/test/java/org/junitpioneer/jupiter/SystemPropertyExtensionTests.java:
```diff
@@ -0,0 +1,95 @@
+		@Test
+		@DisplayName("should clear system property")
+		void shouldClearSystemProperty() {
+			assertThat(System.getProperty("some property")).isNull();
```
A little over the top, maybe, but...

```java
assertThat(System.getProperty("another property")).isEqualTo("old value")
```

?

nicolaiparlog commented on src/test/java/org/junitpioneer/jupiter/SystemPropertyExtensionTests.java:
```diff
@@ -0,0 +1,95 @@
+		void clearAndSetSystemPropertyShouldBeMixable() {
+			assertThat(System.getProperty("some property")).isNull();
+			assertThat(System.getProperty("another property")).isEqualTo("new value");
+		}
```
Good thing, this test is here. Gave the idea to check what happens in this case:

```java
@ClearSystemProperty(key = "some property")
@SetSystemProperty(key = "some property", value = "new value")
```

I'd consider this an implementation error and think the extension should throw an exception to make the developer aware of it by throwing an exception.

nicolaiparlog commented on docs/system-properties.adoc:
```diff
@@ -0,0 +1,42 @@
+
+The `@ClearSystemProperty` and `@SetSystemProperty` annotations can be used to clear, respectively, set the values of system properties for a test execution.
+Both annotations work on the test method level and are repeatable as well as combinable.
+After the annotated method has been executed, the initial default value is restored.
```
See comments below - we may have to update these lines.

beatngu13 commented:
Thanks for your review, very helpful comments! Resolved all but:

* https://github.com/junit-pioneer/junit-pioneer/pull/133#discussion_r356209953
* https://github.com/junit-pioneer/junit-pioneer/pull/133#discussion_r356213963

Have to think about them … 🤔

nicolaiparlog commented:
Hey @beatngu13, looking very good. I got sidetracked into refactoring storing/restoring the backup properties ~> 45b54ad. I wasn't quite happy how the code (almost) repeated itself when storing and when restoring, but I'm not sure, whether my refactoring improved matters much (or at all).

Have a look and let me know what you think - we can revert it if you want. Otherwise we're done here. :raised_hands: Time to come up with [a commit message](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#merging).

Bukama commented:
> Hey @beatngu13, looking very good. I got sidetracked into refactoring storing/restoring the backup properties ~> [45b54ad](https://github.com/junit-pioneer/junit-pioneer/commit/45b54adcddf75519813a1c4ef2801d881d0b6224). I wasn't quite happy how the code (almost) repeated itself when storing and when restoring, but I'm not sure, whether my refactoring improved matters much (or at all).
> 
> Have a look and let me know what you think - we can revert it if you want. Otherwise we're done here. 🙌 Time to come up with [a commit message](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#merging).

As written in chat yesterday I like the refactoring as it's reduce duplicate code and improves readabilty of some things (like the check for properties that were `@Set` and `@Clear`ed). And, you introduced a check for duplicate annotations - even if we have to wait what will come out of [Junit5 Issue 2131](https://github.com/junit-team/junit5/issues/2131).

beatngu13 commented:
Hey @nicolaiparlog @Bukama,

I wanted to join the Twitch stream, but our daughter was born earlier than we expected … 😉

However, I had a quick look at the changes and they are looking really good! I also didn't like the redundancy I had in my code, so thanks for taking the time and fixing it. Also the fact that we now check for duplicate annotations is really helpful.

Will go over the changes in detail and create a proper commit message ASAP.

verhas commented on src/main/java/org/junitpioneer/jupiter/SystemPropertyExtension.java:
```diff
@@ -0,0 +1,167 @@
+		Map<String, String> propertiesToSet;
+		try {
+			propertiesToClear = findRepeatableAnnotations(context, ClearSystemProperty.class).stream()
+					// collect to map because the collector throws an IllegalStateException on
```
Generally: if there is a need for an inline comment then it is a code smell. In this case, the code uses a `Map` instead of a `Set` because the collector to the `Map` throws an exception, while the collector to the `Set` does not (which is btw. absolutely logical).

The nice solution is to use a special set that throws an exception when an element already in the set is added:

```java
private static class UniqueThrowingSet extends HashSet {
        @Override
        public boolean add(Object o) {
            if (contains(o)) {
                throw new IllegalStateException("Duplicate element cannot be added");
            }
            return super.add(o);
        }
    }
```
I am not advocating this solution against the current solution, because this seems to be a bit overkill for the actual problem.

beatngu13 commented on src/main/java/org/junitpioneer/jupiter/SystemPropertyExtension.java:
```diff
@@ -73,68 +97,71 @@ private void handleSystemProperties(ExtensionContext context) {
+	private static class SystemPropertyBackup {
+
+		private final Map<String, String> propertiesToSet;
+		private final Collection<String> propertiesToUnset;
```
Wouldn't `Set` be more appropriate?

beatngu13 commented:
Proposed commit message:

```
Implement @ClearSystemProperty and @SetSystemProperty

Sometimes it is helpful to clear and/or set system properties for a
test execution. This PR introduces two annotations,
`@ClearSystemProperty` and `@SetSystemProperty`, that allow to do
this in a save manner with a proper restore mechanism. Both
annotations work on the test method and class level, are repeatable
as well as combinable.

Fixes: #129

PR: #133
```

nicolaiparlog commented:
Thanks a lot, @beatngu13 for the work and perseverance. :+1:

nicolaiparlog commented:
Wow, PR went from 2018 to 2020.
